dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -728,7 +729,19 @@ BufferedFrame& FramesDecoderGpu::FindEmptySlot() {
       return frame;
     }
   }
-  DALI_FAIL("Could not find empty slot in the frame buffer");
+
+  // in some cases we may decode more than one frame after receiving one packet
+  // frame N-1 may require packet N and N-1, and in results after submitting packet N we
+  // will get frame N-1 and N but the buffer may have space only for 1 frame
+  std::vector<BufferedFrame> new_frame_buffer(frame_buffer_.size() + 1);
+  for (size_t i = 0; i < frame_buffer_.size(); ++i) {
+    new_frame_buffer[i] = std::move(frame_buffer_[i]);
+  }
+  frame_buffer_ = std::move(new_frame_buffer);
```

Comment:
I was wondering why you decided to create new buffer and move it in place of the old one instead of `push_back` new frame to the old buffer?

Reply:
Because `DeviceBuffer` doesn't have assignment operator:
```

In file included from /Dali/dali/dali/operators/reader/loader/video/frames_decoder_gpu.h:35,
                 from /Dali/dali/dali/operators/reader/loader/video/frames_decoder_gpu.cc:14:
/Dali/dali/include/dali/core/dev_buffer.h: In instantiation of ‘dali::DeviceBuffer<T>::DeviceBuffer(dali::DeviceBuffer<T>&&) [with T = unsigned char]’:
/Dali/dali/dali/operators/reader/loader/video/frames_decoder_gpu.h:123:8:   required from ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = dali::BufferedFrame; _Args = {dali::BufferedFrame}; _Tp = dali::BufferedFrame]’
/usr/include/c++/11/bits/alloc_traits.h:516:17:   required from ‘static void std::allocator_traits<std::allocator<_Tp1> >::construct(std::allocator_traits<std::allocator<_Tp1> >::allocator_type&, _Up*, _Args&& ...) [with _Up = dali::BufferedFrame; _Args = {dali::BufferedFrame}; _Tp = dali::BufferedFrame; std::allocator_traits<std::allocator<_Tp1> >::allocator_type = std::allocator<dali::BufferedFrame>]’
/usr/include/c++/11/bits/vector.tcc:115:30:   required from ‘std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::emplace_back(_Args&& ...) [with _Args = {dali::BufferedFrame}; _Tp = dali::BufferedFrame; _Alloc = std::allocator<dali::BufferedFrame>; std::vector<_Tp, _Alloc>::reference = dali::BufferedFrame&]’
/usr/include/c++/11/bits/stl_vector.h:1204:21:   required from ‘void std::vector<_Tp, _Alloc>::push_back(std::vector<_Tp, _Alloc>::value_type&&) [with _Tp = dali::BufferedFrame; _Alloc = std::allocator<dali::BufferedFrame>; std::vector<_Tp, _Alloc>::value_type = dali::BufferedFrame]’
/Dali/dali/dali/operators/reader/loader/video/frames_decoder_gpu.cc:736:26:   required from here
/Dali/dali/include/dali/core/dev_buffer.h:73:11: error: use of deleted function ‘dali::DeviceBuffer<unsigned char>& dali::DeviceBuffer<unsigned char>::operator=(const dali::DeviceBuffer<unsigned char>&)’
   73 |     *this = other;
      |     ~~~~~~^~~~~~~
/Dali/dali/include/dali/core/dev_buffer.h:70:8: note: ‘dali::DeviceBuffer<unsigned char>& dali::DeviceBuffer<unsigned char>::operator=(const dali::DeviceBuffer<unsigned char>&)’ is implicitly declared as deleted because ‘dali::DeviceBuffer<unsigned char>’ declares a move constructor or move assignment operator
   70 | struct DeviceBuffer {
      |        ^~~~~~~~~~~~
```
No idea why vector quires it and cannot use move operator insteada.